### PR TITLE
Implement `PRE31-C`

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Implements a heuristic query to find cases where arguments are provided to unsafe macros which have side-effects.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - `PRE31-C`
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [x] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: rule_packages/c/SideEffects4.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    "PRE31-C": {
+      "properties": {
+        "obligation": "rule"
+      },
+      "queries": [
+        {
+          "description": "Macro arguments can be expanded multiple times which can cause side-effects to be evaluated multiple times.",
+          "kind": "problem",
+          "name": "Avoid side effects in arguments to unsafe macros",
+          "precision": "low",
```
Comment (knewbury01) [inline rule_packages/c/SideEffects4.json:12]: maybe add implementation note for the details of this precision, ie as discussed in meeting that we do not look for the no use of a parameter case

**Review by knewbury01 — COMMENTED** [review]
just a supplemental few points for the conversation that we had in the team sync!

File: rule_packages/c/SideEffects4.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+  "CERT-C": {
+    "PRE31-C": {
+      "properties": {
+        "obligation": "rule"
+      },
+      "queries": [
+        {
+          "description": "Macro arguments can be expanded multiple times which can cause side-effects to be evaluated multiple times.",
```
[the remark this answers is not in the document]
Comment (lcartey) [reply rule_packages/c/SideEffects4.json:9]: Added.

File: rule_packages/c/SideEffects4.json (hunk 1 above)
Comment (lcartey) [reply rule_packages/c/SideEffects4.json:12]: Done.

File: c/cert/src/rules/PRE31-C/SideEffectsInArgumentsToUnsafeMacros.ql [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import cpp
+import codingstandards.c.cert
+import codingstandards.cpp.Macro
+import codingstandards.cpp.SideEffect
+import codingstandards.cpp.StructuralEquivalence
+import codingstandards.cpp.sideeffect.DefaultEffects
+import codingstandards.cpp.sideeffect.Customizations
+import semmle.code.cpp.valuenumbering.HashCons
+
+class FunctionCallEffect extends GlobalSideEffect::Range {
```
[the remark this answers is not in the document]
Comment (lcartey) [reply c/cert/src/rules/PRE31-C/SideEffectsInArgumentsToUnsafeMacros.ql:26]: Done.

**Pull request merged**